src/uaclient/uasession.cpp

```diff
@@ -0,0 +1,182 @@
+ *  the Free Software Foundation, either version 3 of the Licence.
+ *
+ *  Quasar is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public Licence for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with Quasar.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdexcept>
+
+#include <open62541.h>
+#include <open62541_compat_common.h>
+#include <uaclient/uasession.h>
+#include <uadatavalue.h>
+#include <LogIt.h>
+
+namespace UaClientSdk
+{
+
+UaSession::UaSession():
+        m_client(0)
+{
+
+}
+
+UaStatus UaSession::connect(
+        const UaString&                endpoint,
+        const SessionConnectInfo&      connectInfo,
+        const SessionSecurityInfo&     securityInfo,
+        UaSessionCallback*             callback)
+{
+
+    if (m_client)
+        return OpcUa_Bad;  // Already connected!
+    m_client = UA_Client_new(UA_ClientConfig_standard);
+    if (!m_client)
+        throw alloc_error();
+    UaStatus status = UA_Client_connect(m_client, endpoint.toUtf8().c_str());
+    if(! status.isGood())
+    {
+        UA_Client_delete(m_client);
+        LOG(Log::ERR) << "in UaSession::connect() " << status.toString().toUtf8();
+        return status;
+    }
+    return OpcUa_Good;
+}
+
+/**
+ * What's not implemented:
+ * - diagnostic infos (TODO)
+ * - taking into account serviceSettings (TODO)
+ * - type of timestamps (TODO)
+ * - maxAge (TODO)
+ *
+ * also: this code is highlt non-reentrant because of open62541- TODO: mutex it!
+ */
+UaStatus UaSession::read(
+            ServiceSettings &           serviceSettings,
+            OpcUa_Double                maxAge,
+            OpcUa_TimestampsToReturn    timeStamps,
+            const UaReadValueIds &      nodesToRead,
+            UaDataValues &              values,
+            UaDiagnosticInfos &         diagnosticInfos  )
+{
+    if (nodesToRead.size() != 1)
+    {
+        throw std::runtime_error("So far only single reads are supported"); // FIXME
+    }
+
+    if (nodesToRead.size() != values.size())
+        throw std::runtime_error("Size of provided values must match size of nodesToRead");
+
+    UA_ReadRequest readRequest;
+    UA_ReadRequest_init(&readRequest);
+    readRequest.nodesToRead =  (UA_ReadValueId*) UA_Array_new(1, &UA_TYPES[UA_TYPES_READVALUEID]);
+    readRequest.nodesToReadSize = 1;
+
+    UA_NodeId_init(&readRequest.nodesToRead[0].nodeId);
+    nodesToRead[0].NodeId.copyTo( &readRequest.nodesToRead[0].nodeId );
+
+    readRequest.nodesToRead[0].attributeId = nodesToRead[0].AttributeId;
+
+    UA_ReadResponse readResponse = UA_Client_Service_read(m_client, readRequest);
+
+    UA_Array_delete(readRequest.nodesToRead, 1, &UA_TYPES[UA_TYPES_READVALUEID]);
```

Comment:
Would shared/unique ptr with custom delete work here for readRequest (with UA_Array_delete). Just in case something throws or future developers miss/add an exit path to this function.

Reply:
Here it would probably help. Spawned as OPCUA-937 improvement ticket.